src/Bicep.Core/TypeSystem/Az/AzResourceTypeProvider.cs

```diff
@@ -44,36 +49,135 @@ public AzResourceTypeProvider(ITypeLoader typeLoader)
         }
 
-        public ResourceType GetType(ResourceScope scopeType, ResourceTypeReference typeReference)
+        private ResourceType GenerateResourceType(ResourceTypeReference typeReference, bool isExistingResource)
         {
-            if (loadedTypeCache.TryGetValue(typeReference, out var resourceType))
+            if (availableResourceTypes.TryGetValue(typeReference, out var typeLocation))
             {
-                return resourceType;
+                var serializedResourceType = typeLoader.LoadResourceType(typeLocation);
+                var resourceType = resourceTypeFactory.GetResourceType(serializedResourceType);
+
+                return SetBicepResourceProperties(resourceType, isExistingResource);
             }
+            else
+            {
+                var resourceBodyType = new NamedObjectType(typeReference.FormatName(), TypeSymbolValidationFlags.Default, LanguageConstants.CreateResourceProperties(typeReference), null);
+                var resourceType = new ResourceType(typeReference, ResourceScope.Tenant | ResourceScope.ManagementGroup | ResourceScope.Subscription | ResourceScope.ResourceGroup | ResourceScope.Resource, resourceBodyType);
 
-            if (availableResourceTypes[scopeType].TryGetValue(typeReference, out var typeLocation))
+                return SetBicepResourceProperties(resourceType, isExistingResource);
+            }
+        }
+
+        public static ResourceType SetBicepResourceProperties(ResourceType resourceType, bool isExistingResource)
+        {
+            var bodyType = resourceType.Body.Type;
+
+            switch (bodyType)
             {
-                // It's important to cache this result because LoadResourceType is an expensive operation, and
-                // duplicating types means the resourceTypeFactor won't be able to use its cache.
-                var serializedResourceType = typeLoader.LoadResourceType(typeLocation);
-                resourceType = resourceTypeFactory.GetResourceType(serializedResourceType);
+                case ObjectType bodyObjectType:
+                    bodyType = SetBicepResourceProperties(bodyObjectType, resourceType.ValidParentScopes, isExistingResource);
+                    break;
+                case DiscriminatedObjectType bodyDiscriminatedType:
+                    var bodyTypes = bodyDiscriminatedType.UnionMembersByKey.Values.ToList().Select(x => x.Type as ObjectType ?? throw new ArgumentException());
+                    bodyTypes = bodyTypes.Select(x => SetBicepResourceProperties(x, resourceType.ValidParentScopes, isExistingResource));
+                    bodyType = new DiscriminatedObjectType(
+                        bodyDiscriminatedType.Name,
+                        bodyDiscriminatedType.ValidationFlags,
+                        bodyDiscriminatedType.DiscriminatorKey,
+                        bodyTypes);
+                    break;
+                default:
+                    throw new ArgumentException();
```

Comment:
>ArgumentException [](start = 30, length = 17)

Should provide a message so we can find it.

Reply:
Will do. We test type deserialization for the full set of types, and it's fully deterministic, so this exception will only be thrown in CI - so I've been lazy about putting messages in.